### PR TITLE
Support configFailurePolicy for TestNG 7.x

### PR DESCRIPTION
**Pull request by DPUkyle**
See #10507 for context.

TestNG 7.x removed the `TestNG#setConfigFailurePolicy(String)` method in favor of `TestNG#setConfigFailurePolicy(org.testng.xml.XmlSuite$FailurePolicy)` (an enum argument).

Older (pre-5.12) TestNG libraries did not have this method at all, so we probe at configuration time to see if the user has configured this property on our `TestNGOptions` bean. So the first step was to update `TestNGTestFramework` to look for the newer API first, then fallback to String, finally failing if neither are present (this "probing" is skipped entirely if the [option](https://docs.gradle.org/7.4/javadoc/org/gradle/api/tasks/testing/testng/TestNGOptions.html#setConfigFailurePolicy-java.lang.String-) is not set/left as default.)

Next, when invoking the TestNG class we must actually invoke the method. Previously we just trusted that the String argument method existed since we've passed the configuration-time probing/validation. But since our spec bean stores a string, we must coerce its value to the enum value, if appropriate. So additional reflection is added to `TestNGTestClassProcessor`. The end result is we invoke the new API if available, otherwise fallback to String.

Lastly, we also probe for marker interfaces in `TestNGListenerAdapterFactory`. I updated this to probe for the newest/non-deprecated item in TestNG 7.x.

@big-guy and I discussed the following alternative approaches below, but ultimately don't want to make the additional investment at this point in time:

1. In `TestNGTestFramework`, avoid leveraging the `testClassLoader` entirely and instead try to extract the TestNG version from the associated dependency. Based on that, we can make assumptions about which APIs are OK to call based on the library version.
2. Build our own API adapter layer. Similar to above, but make our options/spec object smarter about which APIs we invoke.
3. Investigate [https://github.com/testng-team/testng-remote](testng-team/testng-remote). This came up in community discussions as a way to abstract the TestNG version from calls to invoke it.
4. Stop invoking TestNG directly; instead use JUnitPlatform and a custom runner for TestNG: [junit-team/testng-engine](https://github.com/junit-team/testng-engine)

If we decide to pursue further TestNG integration in the future, the above items deserve further exploration.

### CODE REVIEW

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Comment by DPUkyle** [comment]
FYI @juherr @krmahadevan

**Comment by big-guy** [comment]
@bot-gradle test and merge

File: subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/testng/TestNGTestClassProcessor.java [hunk 1]
```diff
@@ -109,7 +110,11 @@ private void runTests() {
         if (options.getThreadCount() > 0) {
             testNg.setThreadCount(options.getThreadCount());
         }
-        invokeVerifiedMethod(testNg, "setConfigFailurePolicy", String.class, options.getConfigFailurePolicy(), DEFAULT_CONFIG_FAILURE_POLICY);
+
+        Class<?> configFailurePolicyArgType = getConfigFailurePolicyArgType(testNg);
+        Object configFailurePolicyArgValue = getConfigFailurePolicyArgValue(testNg);
```
Comment (marcphilipp) [inline subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/testng/TestNGTestClassProcessor.java:115]: Could the type be derived by calling `getClass()` on the value?

**Comment by juherr** [comment]
As I was tagged, my opinion about the alternative approaches :

https://github.com/testng-team/testng-remote is the best current solution from testng-team pov. Mainly because we need it for the eclipse plugin and we provide some support. It always makes me sad when I see so many implementations of the abstraction of the testng api (at least: maven, intellij, eclipse)

https://github.com/junit-team/testng-engine is a good target because it will allow you to manage all engines with one api only but I don't know if it is already production-ready. It is not the case but `testng-engine` could use `testng-remote` under the hood In order to cover the maximum of testng versions (if I remember well it is not possible because `testng-engine` needs apis that don't exist in `testng-remote`)

**Comment by DPUkyle** [comment]
Thanks for the response, Julien!

> https://github.com/testng-team/testng-remote is the best current solution from testng-team pov.

I completely understand as this is your API abstraction layer. We would also prefer not to use reflection. My biggest hesitation is that the testng-remote binaries appear to be published to bintray and are no longer accessible. Do you have plans to move or publish future versions to Maven Central?

**Comment by juherr** [comment]
> Do you have plans to move or publish future versions to Maven Central?

I missed this problem. Sure, we will fix that.

**Pull request merged**